File: Problem1.java
// Time Complexity : O(log(min(m,n)) , where m and n are the length of the array 1 and array2 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no


class Solution {
    // Brute Force Solution
    /*
    public int[] intersect(int[] nums1, int[] nums2) {

        Map<Integer,Integer> map = new HashMap<>(); // element to count map
        List<Integer> list = new ArrayList<>();
        int[] result= new int[0];

        // assuming nums 1 is smaller
        for(int i=0;i<nums1.length;i++){
            if(!map.containsKey(nums1[i])){
                map.put(nums1[i] , 0);
            }
            map.put(nums1[i] , map.get(nums1[i]) + 1);
        }

        for(int m=0;m< nums2.length; m++){
            if(map.containsKey(nums2[m])){
                if(map.get(nums2[m]) > 0 ){
                    list.add(nums2[m]);
                    //reduce the count
                    map.put(nums2[m] , map.get(nums2[m]) - 1);
                }
            }
        }
        
        if(list.size() > 0 ){
            result = new int[list.size()];
            int i=0;
            for(Integer item : list){
                result[i++] = item;
            }
        }

        return result;
    }
    */

    // arrays sorted solution
    /*
    Follow up, when arrays already sorted

    Ignoring the complexity of sorted array
    Time : O(m+n) // if adding the sorting complexity then mlog+nlogn+(m+n) 
    Space: O(1)
        public int[] intersect(int[] nums1, int[] nums2) {
            // Assuming arrays are sorted
            // so sorting arrays
            Arrays.sort(nums1);
            Arrays.sort(nums2);

            List<Integer> result = new ArrayList<>();
            int[] resArr= new int[0];
            
            int p1 = 0, p2=0;

            while(p1<nums1.length && p2 < nums2.length){
                if(nums1[p1] == nums2[p2]){
                    result.add(nums1[p1]);
                    p1++;
                    p2++;
                }
                else if(nums1[p1] < nums2[p2]){
                    p1++;
                }
                else{
                    p2++;
                }

            }

            if(result.size() > 0){
                resArr = new int[result.size()];
                int i=0;
                for(Integer item : result){
                    resArr[i++] = item;
                }
            }

            return resArr;
    }
    */
    // when both arrays are sorted and one's length is larger than other'
    public int[] intersect(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);

         int[] largeArr= new int[0];
         int[] smallArr= new int[0];

         if(nums1.length > nums2.length){
             largeArr = nums1;
             smallArr = nums2;
         }
         else{
             largeArr = nums2;
             smallArr = nums1;
         }

         int globalLow= 0;
         List<Integer> list = new ArrayList<>();
         int[] result = new int[0];
         int lPtr = 0;

         while(lPtr<smallArr.length){
             int idx = binarySearch(largeArr , smallArr[lPtr] , globalLow , largeArr.length - 1);
             if(idx == -1){
                 // it means that its not found
                 lPtr++;
             }
             else{
                 list.add(smallArr[lPtr]);
                 lPtr++;
                 globalLow = idx + 1;
             }             
         }

        if(list.size() > 0){
                result = new int[list.size()];
                int i=0;
                for(Integer item : list){
                    result[i++] = item;
                }
            }

            return result;
    }

    private int binarySearch(int[] arr , int x , int low , int high){
        int mid = low;
        while(low<=high){
            mid = low + (high -low)/2;
            if(arr[mid] == x){
                // check for the first occurance
                if(mid>0 && mid-1 >=low &&  arr[mid-1] == arr[mid]){ // If there is another occurance on the left side, then we will move on left but if have already found the occurance previlusly thant mid-1<low then we will not move left again (edge case)
                    high = mid-1; // go on the left to find the first occurance
                }
                else{
                return mid;
                }
            }
            else if (arr[mid] < x){
                low = mid+1;
            }
            else{
                high = mid - 1;
            }
        }

        return -1;
    }

}

File: problem1.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/*

Store the frequency of the occurences of all the values in array 1
Check the values of array 2 in array 1 and if found reduce the frequency.

*/

// using hashmap

#include<iostream>
#include<unordered_map>
#include<vector>

using namespace std;

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res{};
        unordered_map<int,int> nums1_freq{};
        for(auto& c:nums1){
            if(nums1_freq.find(c) == nums1_freq.end()){
                nums1_freq[c] = 1;
            }
            else{
                nums1_freq[c]++;
            }
        }
        for(auto& c:nums2){
            if(nums1_freq.find(c)!=nums1_freq.end() && nums1_freq.at(c)>0){
                res.push_back(c);
                nums1_freq.at(c)--;
            }
        }
        return res;
    }
};

// Time Complexity : O(NlogN) // sorting + O(NlogN) finding n elements of array 2 in array 1
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/*

Initially initialize the low as 0 and high as length - 1 for search array.
Search the elements of array 2 in array 1 using binary search
If repetition in array 1 then check if the last elemetn in mid-1 is not the target element. If not choose that mid value
and add that to the result. 
Update the new search area for array1 to idx given+1 so that the repeat is eliminated from the search .
Also since the array is sorted it helps avoid repition.
Do untill all elements of array 2 are iterated.

Here array1 is larger than array 2.

*/


// using binary search;

class Solution {
    int bs(vector<int>& nums1,int low,int high,int target){
        int gg_low = low;
        while(low<=high){
            int mid = low + (high - low)/2;
            //cout<<"Target "<<target<<"mid_value "<<mid<<"nums_val "<<nums1.at(mid)<<"low_val "<<low<<endl;
            if(nums1.at(mid) == target){
                if(mid == gg_low) return mid;
                if(nums1.at(mid-1)!= target) return mid;
            }
            if(nums1.at(mid)>=target){
                high = mid - 1;
                //cout<<"change high"<<endl;
            }
            else {
                low = mid + 1;
                //cout<<"change low"<<endl;
            }    
        }
        return -1;
    }

public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res{};
        sort(nums1.begin(),nums1.end());
        sort(nums2.begin(),nums2.end());
        int n1 = nums1.size();
        int n2 = nums2.size();
        if(n2>n1) return intersect(nums2,nums1);
        int low{},high{n1-1};
        for(int i{};i<n2;++i){
            int target = nums2.at(i);
            int idx = bs(nums1,low,high,target);
            if(idx != -1){
                res.push_back(nums1.at(idx));
                low = idx+1;
            }
            cout<<i<<" "<<low<<endl;
        }
        return res;
    }
};